enhancements/network/bgp-ovn-kubernetes.md

```diff
@@ -0,0 +1,701 @@
+
+Another risk is integration with MetalLB. We need to ensure that MetalLB is still able to function correctly with
+whatever OVN-Kubernetes is configuring within FRR.
+
+In addition, by dynamically learning routes and programming them into OVN, we risk a customer accidentally introducing
+potentially hundreds or even thousands of routes into different OVN logical routers. As a mitigation, we can recommend
+using communities or route aggregation to customers to limit RIB size. We will also need to scale test the effect of
+many routes inside OVN logical routers.
+
+Also, FRR-K8S is not GA yet, and targeted for 4.17.
+
+### Drawbacks
+
+* Increased complexity of our SDN networking solution to support more complex networking. 
+* Increases support complexity due to integration with the a user's provider network.
+* Increases network control plane load on nodes, as they have multiple peer connections and are having to maintain
+a dynamic routing protocol.
+
+### Implementation Details/Notes/Constraints
+
+#### BGP Peering
+
+FRR-K8S shall support both internal BGP (iBGP) and external BGP (eBGP). With iBGP, a full mesh topology is required. This
+can be cumbersome and have issues with scaling in large clusters. In the future, FRR-K8S shall also support designating one
+or more nodes as route reflectors (https://issues.redhat.com/browse/CNF-10719). With route reflectors, nodes only have
+to connect to the route reflectors and do not have to maintain a full mesh with all other nodes in the cluster. Therefore,
+support for iBGP in this enhancement will be restricted to full mesh only.
+
+#### FRR-K8S Integration
+
+As previously mentioned frr-k8s will be used in order to deploy and manage FRR configuration. Support will be added
+where necessary for gaps in frr-k8s API, but in the meantime frr-k8s supports also issuing raw FRR configuration. Every
+attempt should be made to leverage the FRR-K8S API, and raw configuration should only be used as a last resort. A user
+will need the ability to:
+
+* Define BGP peers, ASN, authentication, along with other common BGP configurations like communities, route filtering, etc.
+* Configure BFD
+
+FRR-K8S will serve as the main API for user configuration of BGP via FRR-K8S CRs. OVN-Kubernetes will watch for these
+CRs, and once created, OVN-Kubernetes will start monitoring the kernel routing table via netlink for routes installed
+by FRR. When FRR is using Zebra, the routes can be identified via a proto "bgp" label:
+
+```
+123.123.123.0/24 nhid 47 via 172.18.0.5 dev breth0 proto bgp metric 20
+```
+
+When these routes are seen on the node, ovnkube-controller will update the routes in NBDB for any GW routers on the
+node, filtering the route for the GW router based on if the interface on the route matches the interface in OVS. This
+will happen irrespective of gateway mode, as some features in local gateway mode rely on the GRs for sending traffic
```

Comment:
Local gateway mode exists mostly only to support custom routing at this point, right? Is there any plan to kill it off in favor of BGP?

Reply:
I'd say local gateway is independent with that regard, as users (for example telco ones) might need each node to be connected to multiple networks via multiple interfaces, while adding routes to the host either statically (via knmstate) or dynamically (learning them via bgp), which won't work with shared gw mode as all the egress traffic leaves via br-ex.

Reply:
To get rid of local gw we would need to be able to manage other interfaces on the node with OVN, which means adding them to their own OVS bridges. We would also need KAPIs to program routes in OVN when not using BGP as some customers may not use a dynamic routing protocol. In the near future we dont have any of these things. 

Shared gateway gives us the advantage of offload and performance, while local gw gives us the flexibility to lean on the kernel for things that OVN doesn't support. I'm not sure if we will ever consolidate to just one.